### DISCUSSION TITLE
How to send Embed

### DISCUSSION THREAD

**Discussion by N0m4n904** [kordlib/kord#939] [Help] [answered]
Hi, I have this EmbedBuilder:
```
fun buildNormalEmbed(
    buildData: BuildData?,
    color: Color,
    thumbnail: EmbedBuilder.Thumbnail,
    description: String
){
    val embedBuilder = EmbedBuilder()
    embedBuilder.author?.name = "Build 2"
    embedBuilder.author?.url = "https://halogenos.org"
    embedBuilder.author?.icon = null
    embedBuilder.color = color
    embedBuilder.thumbnail = thumbnail
    embedBuilder.description = description
}
```
How do I send this now to a specific channel?

Or am I doing it completely wrong? The documentation is not helping here.

With Discord4j I am doing it like this:

```
fun sendEmbedMessage(embed: EmbedCreateSpec, channelId: String) = reportBotError {
        gateway.getChannelById(Snowflake.of(channelId))
            .ofType(GuildMessageChannel::class.java)
            .flatMap { channel -> channel.createMessage(embed) }
            .subscribe()
    }
```
But I can't seem to get it to work on Kord.

**Comment by lukellmann** [comment, marked as answer]
With kord, you don't have to create builder instances yourself (i.e. calling the `EmbedBuilder` constructor in your case). Instead we have [DSL](https://kotlinlang.org/docs/type-safe-builders.html)-like functions.

Here you would call `channel.createEmbed` which takes a lambda which gets a `EmbedBuilder` passed as its [receiver](https://kotlinlang.org/docs/lambdas.html#function-literals-with-receiver). So the final code would look something like this:
```kotlin
channel.createEmbed {
    color = TODO()
    description = "..."
    author {
        name = "..."
        url = "..."
        // ...
    }
    thumbnail {
        // ...
    }
    // ...
}
```

If you want to extract the logic of filling the embed out of the lambda, you can make use of extension functions:
```kotlin
fun EmbedBuilder.buildNormalEmbed(
    color: Color,
    description: String,
    // ...
) {
    this.color = color
    this.description = description
    author {
        // ...
    }
    thumbnail {
        // ...
    }
}

// ...
channel.createEmbed {
    buildNormalEmbed(...)
}
```

**Reply by N0m4n904** [reply]
Ah okay, and where do I have to set the channelId? Looking at the code it returns a createMessage, but it seems not to have a parameter for the channelId.
I could be wrong though and simply can't find it again,

**Reply by lukellmann** [reply]
You need a `MessageChannel` as the `channel` object in the code above. You could get one by its ID e.g. like this:
```kotlin
val channel = kord.getChannelOf<MessageChannel>(channelId)
```

**Reply by N0m4n904** [reply]
thanks, will try that this evening.

**Reply by N0m4n904** [reply]
Okay, it seems to work, but I have one more question.
I have this Map:
```
val buildMessages = mutableMapOf<Int, Message>()
```

on discord4j I can use this code to store a message in that list inside the send function:

```
fun sendBuildEmbed(buildData: BuildData, embed: EmbedCreateSpec, channelId: String) =
        reportBotError {
            gateway.getChannelById(Snowflake.of(channelId))
                .ofType(GuildMessageChannel::class.java)
                .flatMap { channel -> channel.createMessage(embed) }
                .map { message -> buildMessages[buildData.buildNumber] = message }
                .subscribe()
        }
```

is it possible to do this in kord?

My function right now looks like this:

```
suspend fun sendBuildEmbed(buildData: BuildData, embedBuilder: EmbedBuilder, channelId: String) {
        val channel = sharedKord?.getChannelOf<MessageChannel>(Snowflake(channelId))

        channel?.createEmbed {
            embedBuilder
        }
    }
```

so, am I able to get the message object out of this code to store it inside the map?

**Reply by lukellmann** [reply]
`createEmbed` returns the newly created message, you can simply use it like any function return value

**Reply by N0m4n904** [reply]
Thanks!
